components/brave_ads/content/browser/search_result_ad/search_result_ad_redirect_throttle.cc

```diff
@@ -0,0 +1,106 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/content/browser/search_result_ad/search_result_ad_redirect_throttle.h"
+
+#include <utility>
+
+#include "brave/components/brave_ads/common/search_result_ad_util.h"
+#include "brave/components/brave_ads/content/browser/search_result_ad/search_result_ad_service.h"
+#include "brave/components/brave_search/common/brave_search_utils.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "services/network/public/cpp/request_mode.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "third_party/blink/public/mojom/loader/resource_load_info.mojom-shared.h"
+#include "url/gurl.h"
+
+namespace brave_ads {
+
+namespace {
+
+SessionID GetTabId(content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  SessionID tab_id = sessions::SessionTabHelper::IdForTab(web_contents);
+  content::WebContents* original_web_contents =
+      web_contents->GetFirstWebContentsInLiveOriginalOpenerChain();
+  if (original_web_contents) {
+    tab_id = sessions::SessionTabHelper::IdForTab(original_web_contents);
+  }
+
+  return tab_id;
+}
+
+}  // namespace
+
+// static
+std::unique_ptr<SearchResultAdRedirectThrottle>
+SearchResultAdRedirectThrottle::MaybeCreateThrottleFor(
+    SearchResultAdService* search_result_ad_service,
+    const network::ResourceRequest& request,
+    content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  if (!search_result_ad_service || !request.request_initiator ||
+      !request.has_user_gesture || !request.is_outermost_main_frame ||
```

Comment:
why are you checking user_gesture here?

Reply:
also if we only care about the main frame, why are you using `URLLoaderThrottle` instead of `NavigationThrottle`?

Reply:
> why are you checking user_gesture here?
I thought there was such requirement. Will double check it.

> also if we only care about the main frame, why are you using URLLoaderThrottle instead of NavigationThrottle?
Yes, `NavigationThrottle` is also possible here. I will change